src/mpio.jl

```diff
@@ -0,0 +1,94 @@
+using .MPI
+import Libdl
+
+# Check whether the HDF5 libraries were compiled with parallel support.
+HAS_PARALLEL[] = Libdl.dlopen(libhdf5) do lib
+    Libdl.dlsym(lib, :H5Pget_fapl_mpio, throw_error=false) !== nothing
+end
+
+# Low-level MPI handles.
+const MPIHandle = Union{MPI.MPI_Comm, MPI.MPI_Info}
+
+# MPI.jl wrapper types.
+const MPIHandleWrapper = Union{MPI.Comm, MPI.Info}
+
+const H5MPIHandle = let csize = sizeof(MPI.MPI_Comm)
+    @assert csize in (4, 8)
+    csize == 4 ? Hmpih32 : Hmpih64
+end
+
+h5_to_mpi(handle::H5MPIHandle) = reinterpret(MPI.MPI_Comm, handle)
+
+mpi_to_h5(handle::MPIHandle) = reinterpret(H5MPIHandle, handle)
+mpi_to_h5(mpiobj::MPIHandleWrapper) = mpi_to_h5(mpiobj.val)
+
+# Set MPIO properties in HDF5.
+# Note: HDF5 creates a COPY of the comm and info objects.
+function h5p_set_fapl_mpio(fapl_id, comm::MPI.Comm, info::MPI.Info)
+    h5comm = mpi_to_h5(comm)
+    h5info = mpi_to_h5(info)
+    h5p_set_fapl_mpio(fapl_id, h5comm, h5info)
+end
+
+h5p_set_fapl_mpio(fapl_id, comm::Hmpih32, info::Hmpih32) =
+    h5p_set_fapl_mpio32(fapl_id, comm, info)
+h5p_set_fapl_mpio(fapl_id, comm::Hmpih64, info::Hmpih64) =
+    h5p_set_fapl_mpio64(fapl_id, comm, info)
+
+# Retrieves the copies of the comm and info MPIO objects from the HDF5 property list.
+function h5p_get_fapl_mpio(fapl_id)
+    h5comm, h5info = h5p_get_fapl_mpio(fapl_id, H5MPIHandle)
+    comm = MPI.Comm(h5_to_mpi(h5comm))
+    info = MPI.Info(h5_to_mpi(h5info))
+    comm, info
+end
+
+function h5p_get_fapl_mpio(fapl_id, ::Type{h}) where {h}
+    comm, info = Ref{h}(), Ref{h}()
+    h5p_get_fapl_mpio(fapl_id, comm, info)
+    comm[], info[]
+end
+
+h5p_get_fapl_mpio(fapl_id, comm::Ref{Hmpih32}, info::Ref{Hmpih32}) =
+    h5p_get_fapl_mpio32(fapl_id, comm, info)
+h5p_get_fapl_mpio(fapl_id, comm::Ref{Hmpih64}, info::Ref{Hmpih64}) =
+    h5p_get_fapl_mpio64(fapl_id, comm, info)
+
+hdf5_prop_get_set[:fapl_mpio] =
+    (h5p_get_fapl_mpio, h5p_set_fapl_mpio, H5P_FILE_ACCESS)
+
+hdf5_prop_get_set[:dxpl_mpio] =
+    (h5p_get_dxpl_mpio, h5p_set_dxpl_mpio, H5P_DATASET_XFER)
+
+function check_hdf5_parallel()
+    has_parallel() && return
+    error(
+        "HDF5.jl has no parallel support." *
+        " Make sure that you're using MPI-enabled HDF5 libraries, and that" *
+        " MPI was loaded before HDF5." *
+        " See HDF5.jl docs for details."
+    )
+end
+
+"""
+    h5open(filename, [mode="r"], comm::MPI.Comm, [info::MPI.Info]; pv...)
+
+Open or create a parallel HDF5 file using the MPI-IO driver.
+
+Equivalent to `h5open(filename, mode; fapl_mpio=(comm, info), pv...)`.
+Throws an informative error if the loaded HDF5 libraries do not include parallel
+support.
+
+See the [HDF5 docs](https://portal.hdfgroup.org/display/HDF5/H5P_SET_FAPL_MPIO)
+for details on the `comm` and `info` arguments.
+"""
+function h5open(
+        filename::AbstractString, mode::AbstractString,
+        comm::MPI.Comm, info::MPI.Info = MPI.Info(); pv...
```

Comment:
I'm thinking these should actually be keyword arguments.  Would `driver` be a more appropriate name for `info` ?

Reply:
I guess you're thinking about the h5py interface. Note that the `driver` argument in h5py is different from the `info` argument here, which [allows to pass "hints"](https://portal.hdfgroup.org/display/knowledge/How+to+pass+hints+to+MPI+from+HDF5) to the MPI implementation. This can be very important for I/O performance in large computing clusters. I'm a bit surprised that h5py doesn't support this (or maybe it does, but it's not mentioned in the docs).

In my opinion, having a `driver` argument is redundant. If the user passes MPI objects to `h5open`, then it should be clear that they want to perform parallel I/O. I also like the fact that we can take advantage of multiple dispatch to provide a `h5open` variant taking MPI objects, without the need for keyword arguments.

Reply:
Yeah I was. 

Regarding the arguments, should these be keyword arguments instead of positional. I feel like they should be.

Reply:
In my opinion they should be positional arguments, but I guess that's just personal preference. I'd say it's more natural in Julia, as it takes advantage of multiple dispatch to enable different functionalities.

More practically, an advantage of using positional arguments is that Julia doesn't perform dispatch based on keyword arguments. Therefore, if we were to provide a `h5open(args...; comm, info, etc...)` method, then the new keyword arguments `comm` and `info` would need to be added to the main `h5open` variants in `src/HDF5.jl`. These variants are not aware of MPI, so we cannot set e.g. `info = MPI.Info()` as default value, and we would need to set default values such as `comm = nothing, info = nothing`. In the end, I think this would just increase complexity and visual noise in the main HDF5.jl file, and would go against the idea of putting all MPI-related code in a separate file.

Reply:
Thanks for the clarification. So the main argument for keeping them positional,  is that default values can be passed with values that depend on `MPI`  as their defaults, and we obviously can't have these as default values if we made them keyword arguments.

Reply:
I guess keyword based dispatch (a la https://github.com/simonbyrne/KeywordDispatch.jl )  would help since that would free the keywords `comm, info` since they would dispatch on `MPI` types, but it also seems perhaps out of scope.